test-data/unit/check-selftype.test

```diff
@@ -248,3 +248,46 @@ class A:
 class B(A):
     def __init__(self, arg: T) -> None:
         super(B, self).__init__()
+
+[case testSelfTypeNonsensical]
+# flags: --hide-error-context
+from typing import TypeVar, Type
+
+T = TypeVar('T', bound=str)
+class A:
+    def foo(self: T) -> T:   # E: The type of self 'builtins.str' is not a supertype its class '__main__.A'
+        return self
+
+    @classmethod
+    def cfoo(cls: Type[T]) -> T:  # E: The type of self 'Type[builtins.str]' is not a supertype its class 'Type[__main__.A]'
+        return cls()
+
+Q = TypeVar('Q', bound='B')
+class B:
+    def foo(self: Q) -> Q:
+        return self
+
+    @classmethod
+    def cfoo(cls: Type[Q]) -> Q:
+        return cls()
+    
+class C:
+    def foo(self: C) -> C: return self
+
+    @classmethod
+    def cfoo(cls: Type[C]) -> C:
+        return cls()
+
+class D:
+    def foo(self: str) -> str:  # E: The type of self 'builtins.str' is not a supertype its class '__main__.D'
```

Comment:
Here's a repro for the first mystery; it's indeed related to a lambda:
```python
from typing import Callable
class C:
    @classmethod
    def foo(cls,
            arg: Callable[[int], str] = lambda a: ''
            ):
        pass
```
This gives an error on the 'def' line:
```
__tmp__.py: note: In member "foo" of class "C":
__tmp__.py:4: error: The type of self 'builtins.int' is not a supertype its class '__tmp__.C'
```